### DISCUSSION TITLE
How do I use Bevy Mod Picking on glTF objects?

### DISCUSSION THREAD

**Discussion by hintron** [aevyrie/bevy_mod_picking#163] [Q&A] [answered]
I'm trying to get Bevy Mod Picking to work on glTF objects, like so:

```rust
    // Add in custom object from Blender
    commands
        .spawn_bundle(SceneBundle {
            scene: asset_server.load("gltf/test-box-binary/box.glb#Scene0"),
            transform: Transform::from_xyz(2.0, 0.5, -4.0),
            ..default()
        })
        .insert_bundle(PickableBundle::default());
```
However, this doesn't work. The object doesn't react to hover or click, and the green debug cursor does not show up on hover.

What do I need to do to get a glTF object working with Bevy Mod Picker? How can I further debug this? How did you get those complex monkey objects working? Any help is appreciated. Thanks!

**Comment by aevyrie** [comment, marked as answer]
You need to mark the actual meshes as pickable. You can do this by querying for meshes that are children of the gltf scene, and marking those entities with the applicable picking components.

This will be made easier, and with an actual example, once the refactor is complete. See here: https://github.com/aevyrie/bevy_mod_picking/blob/fe3135648d4d33c2c9daf7d9afb0df520a980d4f/examples/gltf.rs

**Reply by Newbytee** [reply]
This example doesn't seem to apply to the latest version of bevy_mod_picking (0.12). Is there any updated example?

**Reply by aevyrie** [reply]
This has been released in 0.13. Note that the code that marks all children is still applicable to 0,12.

**Comment by PatrickChodowski** [comment]
Hello,

I just wanted to share the solution for this problem that I came up with. Maybe it can help someone

To each gltf scene object I want to be selectable I add flag component:

```
#[derive(Component)]
pub struct UnpickableGLTF;

// Loading scene
if let Some(gltf) = assets_gltf.get(&assets.data["house1"]) {
    commands.spawn(SceneBundle {
        scene: gltf.scenes[0].clone(),
        ..Default::default()
    }).insert(Name::new("House 1")).insert(UnpickableGLTF);
  }

```

Then I run a system that would search for entities with this component and recursively go through their children entities (scene seems to have N levels of nested children entities) and insert PickableBundle::default() to each child with mesh


Solution for recursive query I found here:

https://github.com/bevyengine/bevy/issues/838#issuecomment-772082427

```

fn set_pickible_recursive(
  commands: &mut Commands,
  entity: &Entity,
  mesh_query: &Query<(Entity, &Parent), With<Handle<Mesh>>>,
  children_query: &Query<&Children>,
) {
  for (mesh_entity, mesh_parent) in mesh_query.iter(){
    if mesh_parent.get() == *entity {
      commands.entity(mesh_entity).insert(PickableBundle::default());
    }
  }

  if let Ok(children) = children_query.get(*entity) {
    for child in children.iter() {
      set_pickible_recursive(commands, child, mesh_query, children_query);
    }
  }
}


// system
fn make_gltf_scene_pickable(
                      mut commands: Commands,
                      mut unpickable_query: Query<(Entity, &Name, &Children), With<UnpickableGLTF>>,  
                      mesh_query: Query<(Entity, &Parent), With<Handle<Mesh>>>,
                      children_query: Query<&Children>
                    ){

  for (entity, name, children) in unpickable_query.iter_mut(){
    println!(" [MODELS] Setting Pickable on {name}");
    set_pickible_recursive(&mut commands, &entity, &mesh_query, &children_query);
    commands.entity(entity).remove::<UnpickableGLTF>();
  }

}


```

I hope it can help someone somewhere
